src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp

```diff
@@ -403,4 +406,539 @@ QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingStringWidgetWrapper::c
+      }
+      double min = -999999999;
+      if ( !qgsDoubleNear( numberDef->minimum(), std::numeric_limits<double>::lowest() ) )
+      {
+        min = numberDef->minimum();
+      }
+      if ( mDoubleSpinBox )
+      {
+        mDoubleSpinBox->setMinimum( min );
+        mDoubleSpinBox->setMaximum( max );
+      }
+      else
+      {
+        mSpinBox->setMinimum( static_cast< int >( min ) );
+        mSpinBox->setMaximum( static_cast< int >( max ) );
+      }
+
+      if ( numberDef->flags() & QgsProcessingParameterDefinition::FlagOptional )
+      {
+        mAllowingNull = true;
+        if ( mDoubleSpinBox )
+        {
+          mDoubleSpinBox->setShowClearButton( true );
+          const double min = mDoubleSpinBox->minimum() - 1;
+          mDoubleSpinBox->setMinimum( min );
+          mDoubleSpinBox->setValue( min );
+        }
+        else
+        {
+          mSpinBox->setShowClearButton( true );
+          const int min = mSpinBox->minimum() - 1;
+          mSpinBox->setMinimum( min );
+          mSpinBox->setValue( min );
+        }
+        spinBox->setSpecialValueText( tr( "Not set" ) );
+      }
+      else
+      {
+        if ( numberDef->defaultValue().isValid() )
+        {
+          // if default value for parameter, we clear to that
+          bool ok = false;
+          if ( mDoubleSpinBox )
+          {
+            double defaultVal = numberDef->defaultValue().toDouble( &ok );
+            if ( ok )
+              mDoubleSpinBox->setClearValue( defaultVal );
+          }
+          else
+          {
+            int intVal = numberDef->defaultValue().toInt( &ok );
+            if ( ok )
+              mSpinBox->setClearValue( intVal );
+          }
+        }
+        else if ( !qgsDoubleNear( numberDef->minimum(), std::numeric_limits<double>::lowest() ) )
+        {
+          // otherwise we clear to the minimum, if it's set
+          if ( mDoubleSpinBox )
+            mDoubleSpinBox->setClearValue( numberDef->minimum() );
+          else
+            mSpinBox->setClearValue( static_cast< int >( numberDef->minimum() ) );
+        }
+        else
+        {
+          // last resort, we clear to 0
+          if ( mDoubleSpinBox )
+          {
+            mDoubleSpinBox->setValue( 0 );
+            mDoubleSpinBox->setClearValue( 0 );
+          }
+          else
+          {
+            mSpinBox->setValue( 0 );
+            mSpinBox->setClearValue( 0 );
+          }
+        }
+      }
+
+      if ( mDoubleSpinBox )
+        connect( mDoubleSpinBox, qgis::overload<double>::of( &QgsDoubleSpinBox::valueChanged ), this, [ = ] { emit widgetValueHasChanged( this ); } );
+      else if ( mSpinBox )
+        connect( mSpinBox, qgis::overload<int>::of( &QgsSpinBox::valueChanged ), this, [ = ] { emit widgetValueHasChanged( this ); } );
+
+      return spinBox;
+    };
+  }
+  return nullptr;
+}
+
+void QgsProcessingNumericWidgetWrapper::setWidgetValue( const QVariant &value, QgsProcessingContext &context )
```

Comment:
`context` cannot be const in this case?

Reply:
Nope - the temporary layer store within context may be used here to store map layers loaded for parameter resolution only.

Reply:
parameterAsCrs, used by QgsProcessingCrsWidgetWrapper

Reply:
And just to expand on that answer - when other parameter widgets are ported in future they'll also need non-const access to a context, e.g. the various layer input widgets will need this

Reply:
OK :)